TiFlash crash when using memory tracker

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```
[2022/11/14 01:15:15.192 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 3163) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:406] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:414] ["Access: read."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Unknown si_code."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.201 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x5544c01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+89410561]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fb5c6568630\t<unknown symbol> [libpthread.so.0+63024]\n       0x161a5c6\tMemoryTracker::free(long) [tiflash+23176646]\n                \tdbms/src/Common/MemoryTracker.cpp:176\n       0x15cb8b2\tstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+22853810]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315\n       0x61c614e\tstd::__1::__vector_base<std::__1::shared_ptr<DB::TrackedMppDataPacket>, std::__1::allocator<std::__1::shared_ptr<DB::TrackedMppDataPacket> > >::~__vector_base() [tiflash+102523214]\n                \t/usr/local/bin/../include/c++/v1/vector:487\n       0x15dd622\tDB::HashPartitionWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::partitionAndEncodeThenWriteBlocks() [tiflash+22926882]\n                \tdbms/src/Flash/Mpp/HashPartitionWriter.cpp:122\n       0x618c222\tDB::ExchangeSenderBlockInputStream::readImpl() [tiflash+102285858]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.cpp:43\n       0x556f175\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+89583989]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75\n       0x556ee65\tDB::IProfilingBlockInputStream::read() [tiflash+89583205]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43\n       0x61929de\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+102312414]\n                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270\n       0x61924f6\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+102311158]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x16e8dfb\tDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+24022523]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x16ec343\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+24036163]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x16ec016\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+24035350]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x16ed5c2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+24040898]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fb5c6560ea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 4295) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:406] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:414] ["Access: read."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Unknown si_code."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.205 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x5544c01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+89410561]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fb5c6568630\t<unknown symbol> [libpthread.so.0+63024]\n       0x161a5c6\tMemoryTracker::free(long) [tiflash+23176646]\n                \tdbms/src/Common/MemoryTracker.cpp:176\n       0x15cb8b2\tstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+22853810]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315\n       0x61c614e\tstd::__1::__vector_base<std::__1::shared_ptr<DB::TrackedMppDataPacket>, std::__1::allocator<std::__1::shared_ptr<DB::TrackedMppDataPacket> > >::~__vector_base() [tiflash+102523214]\n                \t/usr/local/bin/../include/c++/v1/vector:487\n       0x15dd622\tDB::HashPartitionWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::partitionAndEncodeThenWriteBlocks() [tiflash+22926882]\n                \tdbms/src/Flash/Mpp/HashPartitionWriter.cpp:122\n       0x618c222\tDB::ExchangeSenderBlockInputStream::readImpl() [tiflash+102285858]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.cpp:43\n       0x556f175\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+89583989]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75\n       0x556ee65\tDB::IProfilingBlockInputStream::read() [tiflash+89583205]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43\n       0x61929de\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+102312414]\n                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270\n       0x61924f6\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+102311158]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x16e8dfb\tDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+24022523]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x16ec343\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+24036163]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x16ec016\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+24035350]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x16ed5c2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+24040898]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fb5c6560ea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.206 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.206 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 5074) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.206 +08:00] [ERROR] [BaseDaemon.cpp:406] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.207 +08:00] [ERROR] [BaseDaemon.cpp:414] ["Access: read."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.207 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Unknown si_code."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.207 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x5544c01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+89410561]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fb5c6568630\t<unknown symbol> [libpthread.so.0+63024]\n       0x161a5c6\tMemoryTracker::free(long) [tiflash+23176646]\n                \tdbms/src/Common/MemoryTracker.cpp:176\n       0x15cb8b2\tstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+22853810]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315\n       0x61c614e\tstd::__1::__vector_base<std::__1::shared_ptr<DB::TrackedMppDataPacket>, std::__1::allocator<std::__1::shared_ptr<DB::TrackedMppDataPacket> > >::~__vector_base() [tiflash+102523214]\n                \t/usr/local/bin/../include/c++/v1/vector:487\n       0x15dd622\tDB::HashPartitionWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::partitionAndEncodeThenWriteBlocks() [tiflash+22926882]\n                \tdbms/src/Flash/Mpp/HashPartitionWriter.cpp:122\n       0x618c222\tDB::ExchangeSenderBlockInputStream::readImpl() [tiflash+102285858]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.cpp:43\n       0x556f175\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+89583989]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75\n       0x556ee65\tDB::IProfilingBlockInputStream::read() [tiflash+89583205]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43\n       0x61929de\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+102312414]\n                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270\n       0x61924f6\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+102311158]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x16e8dfb\tDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+24022523]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x16ec343\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+24036163]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x16ec016\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+24035350]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x16ed5c2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+24040898]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fb5c6560ea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=5466]
```
### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
master 6b393836ea60e5695692e59644e2f9f6aaa81c07

--- a/dbms/src/Flash/Coprocessor/tests/gtest_ti_remote_block_inputstream.cpp
+++ b/dbms/src/Flash/Coprocessor/tests/gtest_ti_remote_block_inputstream.cpp
@@ -59,13 +59,13 @@ struct MockWriter
         return summary;
     }
 
-    void partitionWrite(const TrackedMppDataPacketPtr &, uint16_t) { FAIL() << "cannot reach here."; }
-    void broadcastOrPassThroughWrite(const TrackedMppDataPacketPtr & packet)
+    void partitionWrite(TrackedMppDataPacketPtr &&, uint16_t) { FAIL() << "cannot reach here."; }
+    void broadcastOrPassThroughWrite(TrackedMppDataPacketPtr && packet)
     {
         ++total_packets;
         if (!packet->packet.chunks().empty())
             total_bytes += packet->packet.ByteSizeLong();
-        queue->push(packet);
+        queue->push(std::move(packet));
     }
     void write(tipb::SelectResponse & response)
     {

--- a/dbms/src/Flash/Mpp/BroadcastOrPassThroughWriter.cpp
+++ b/dbms/src/Flash/Mpp/BroadcastOrPassThroughWriter.cpp
@@ -92,7 +92,7 @@ void BroadcastOrPassThroughWriter<ExchangeWriterPtr>::encodeThenWriteBlocks()
     }
     assert(blocks.empty());
     rows_in_blocks = 0;
-    writer->broadcastOrPassThroughWrite(tracked_packet);
+    writer->broadcastOrPassThroughWrite(std::move(tracked_packet));
 }
 
 template class BroadcastOrPassThroughWriter<MPPTunnelSetPtr>;

--- a/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp
+++ b/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp
@@ -176,14 +176,14 @@ void FineGrainedShuffleWriter<ExchangeWriterPtr>::batchWriteFineGrainedShuffle()
 }
 
 template <class ExchangeWriterPtr>
-void FineGrainedShuffleWriter<ExchangeWriterPtr>::writePackets(const TrackedMppDataPacketPtrs & packets)
+void FineGrainedShuffleWriter<ExchangeWriterPtr>::writePackets(TrackedMppDataPacketPtrs & packets)
 {
     for (size_t part_id = 0; part_id < packets.size(); ++part_id)
     {
-        const auto & packet = packets[part_id];
+        auto & packet = packets[part_id];
         assert(packet);
         if (likely(packet->getPacket().chunks_size() > 0))
-            writer->partitionWrite(packet, part_id);
+            writer->partitionWrite(std::move(packet), part_id);
     }
 }
 

--- a/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.h
+++ b/dbms/src/Flash/Mpp/FineGrainedShuffleWriter.h
@@ -42,7 +42,7 @@ public:
 private:
     void batchWriteFineGrainedShuffle();
 
-    void writePackets(const TrackedMppDataPacketPtrs & packets);
+    void writePackets(TrackedMppDataPacketPtrs & packets);
 
     void initScatterColumns();
 

--- a/dbms/src/Flash/Mpp/HashPartitionWriter.cpp
+++ b/dbms/src/Flash/Mpp/HashPartitionWriter.cpp
@@ -122,14 +122,14 @@ void HashPartitionWriter<ExchangeWriterPtr>::partitionAndEncodeThenWriteBlocks()
 }
 
 template <class ExchangeWriterPtr>
-void HashPartitionWriter<ExchangeWriterPtr>::writePackets(const TrackedMppDataPacketPtrs & packets)
+void HashPartitionWriter<ExchangeWriterPtr>::writePackets(TrackedMppDataPacketPtrs & packets)
 {
     for (size_t part_id = 0; part_id < packets.size(); ++part_id)
     {
-        const auto & packet = packets[part_id];
+        auto & packet = packets[part_id];
         assert(packet);
         if (likely(packet->getPacket().chunks_size() > 0))
-            writer->partitionWrite(packet, part_id);
+            writer->partitionWrite(std::move(packet), part_id);
     }
 }
 

--- a/dbms/src/Flash/Mpp/HashPartitionWriter.h
+++ b/dbms/src/Flash/Mpp/HashPartitionWriter.h
@@ -40,7 +40,7 @@ public:
 private:
     void partitionAndEncodeThenWriteBlocks();
 
-    void writePackets(const TrackedMppDataPacketPtrs & packets);
+    void writePackets(TrackedMppDataPacketPtrs & packets);
 
     void sendExecutionSummary();
 

--- a/dbms/src/Flash/Mpp/MPPTunnel.cpp
+++ b/dbms/src/Flash/Mpp/MPPTunnel.cpp
@@ -147,8 +147,7 @@ void MPPTunnel::close(const String & reason, bool wait_sender_finish)
         waitForSenderFinish(false);
 }
 
-// TODO: consider to hold a buffer
-void MPPTunnel::write(const TrackedMppDataPacketPtr & data)
+void MPPTunnel::write(TrackedMppDataPacketPtr && data)
 {
     LOG_TRACE(log, "ready to write");
     {
@@ -158,9 +157,9 @@ void MPPTunnel::write(const TrackedMppDataPacketPtr & data)
             throw Exception(fmt::format("write to tunnel which is already closed."));
     }
 
-    if (tunnel_sender->push(data))
+    auto pushed_data_size = data->getPacket().ByteSizeLong();
+    if (tunnel_sender->push(std::move(data)))
     {
-        auto pushed_data_size = data->getPacket().ByteSizeLong();
         updateMetric(pushed_data_size, mode);
         connection_profile_info.bytes += pushed_data_size;
         connection_profile_info.packets += 1;

--- a/dbms/src/Flash/Mpp/MPPTunnel.h
+++ b/dbms/src/Flash/Mpp/MPPTunnel.h
@@ -72,9 +72,9 @@ public:
     {
     }
 
-    virtual bool push(const TrackedMppDataPacketPtr & data)
+    virtual bool push(TrackedMppDataPacketPtr && data)
     {
-        return send_queue.push(data) == MPMCQueueResult::OK;
+        return send_queue.push(std::move(data)) == MPMCQueueResult::OK;
     }
 
     virtual void cancelWith(const String & reason)
@@ -174,9 +174,9 @@ public:
         , queue(queue_size, func)
     {}
 
-    bool push(const TrackedMppDataPacketPtr & data) override
+    bool push(TrackedMppDataPacketPtr && data) override
     {
-        return queue.push(data);
+        return queue.push(std::move(data));
     }
 
     bool finish() override
@@ -273,7 +273,7 @@ public:
     const String & id() const { return tunnel_id; }
 
     // write a single packet to the tunnel's send queue, it will block if tunnel is not ready.
-    void write(const TrackedMppDataPacketPtr & data);
+    void write(TrackedMppDataPacketPtr && data);
 
     // finish the writing, and wait until the sender finishes.
     void writeDone();

--- a/dbms/src/Flash/Mpp/MPPTunnelSet.cpp
+++ b/dbms/src/Flash/Mpp/MPPTunnelSet.cpp
@@ -56,23 +56,21 @@ void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
 }
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(const TrackedMppDataPacketPtr & packet)
+void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(TrackedMppDataPacketPtr && packet)
 {
     checkPacketSize(packet->getPacket().ByteSizeLong());
     RUNTIME_CHECK(!tunnels.empty());
-    for (auto & tunnel : tunnels)
-    {
-        // We should copy the tracked packet for local tunnel.
-        // Because `switchMemoryTracker` will be called later in `readForLocal`.
-        tunnel->write(tunnel->isLocal() && tunnels.size() > 1 ? packet->copy() : packet);
-    }
+    // TODO avoid copy packet for broadcast.
+    for (size_t i = 1; i < tunnels.size(); ++i)
+        tunnels[i]->write(packet->copy());
+    tunnels[0]->write(std::move(packet));
 }
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::partitionWrite(const TrackedMppDataPacketPtr & packet, int16_t partition_id)
+void MPPTunnelSetBase<Tunnel>::partitionWrite(TrackedMppDataPacketPtr && packet, int16_t partition_id)
 {
     checkPacketSize(packet->getPacket().ByteSizeLong());
-    tunnels[partition_id]->write(packet);
+    tunnels[partition_id]->write(std::move(packet));
 }
 
 template <typename Tunnel>

--- a/dbms/src/Flash/Mpp/MPPTunnelSet.h
+++ b/dbms/src/Flash/Mpp/MPPTunnelSet.h
@@ -48,9 +48,9 @@ public:
     // this is a root mpp writing.
     void write(tipb::SelectResponse & response);
     // this is a broadcast or pass through writing.
-    void broadcastOrPassThroughWrite(const TrackedMppDataPacketPtr & packet);
+    void broadcastOrPassThroughWrite(TrackedMppDataPacketPtr && packet);
     // this is a partition writing.
-    void partitionWrite(const TrackedMppDataPacketPtr & packet, int16_t partition_id);
+    void partitionWrite(TrackedMppDataPacketPtr && packet, int16_t partition_id);
     // this is a execution summary writing.
     void sendExecutionSummary(tipb::SelectResponse & response);
 

--- a/dbms/src/Flash/Mpp/tests/gtest_mpp_exchange_writer.cpp
+++ b/dbms/src/Flash/Mpp/tests/gtest_mpp_exchange_writer.cpp
@@ -120,8 +120,8 @@ struct MockExchangeWriter
         , part_num(part_num_)
     {}
 
-    void broadcastOrPassThroughWrite(const TrackedMppDataPacketPtr & packet) { checker(packet, 0); }
-    void partitionWrite(const TrackedMppDataPacketPtr & packet, uint16_t part_id) { checker(packet, part_id); }
+    void broadcastOrPassThroughWrite(TrackedMppDataPacketPtr && packet) { checker(packet, 0); }
+    void partitionWrite(TrackedMppDataPacketPtr && packet, uint16_t part_id) { checker(packet, part_id); }
     void write(tipb::SelectResponse &) { FAIL() << "cannot reach here, only consider CH Block format"; }
     void sendExecutionSummary(tipb::SelectResponse & response)
     {

--- a/dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp
+++ b/dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp
@@ -27,6 +27,16 @@ namespace DB
 {
 namespace tests
 {
+namespace
+{
+TrackedMppDataPacketPtr newDataPacket(const String & data)
+{
+    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
+    data_packet_ptr->getPacket().set_data(data);
+    return data_packet_ptr;
+}
+} // namespace
+
 class MockPacketWriter : public PacketWriter
 {
     bool write(const mpp::MPPDataPacket & packet) override
@@ -319,10 +329,8 @@ TEST_F(TestMPPTunnel, WriteAfterUnconnectFinished)
     try
     {
         auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
-        auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
         setTunnelFinished(mpp_tunnel_ptr);
-        data_packet_ptr->getPacket().set_data("First");
-        mpp_tunnel_ptr->write(data_packet_ptr);
+        mpp_tunnel_ptr->write(newDataPacket("First"));
         GTEST_FAIL();
     }
     catch (Exception & e)
@@ -353,9 +361,7 @@ try
     auto mpp_tunnel_ptr = constructRemoteSyncTunnel();
     mpp_tunnel_ptr->connect(writer_ptr.get());
     GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->close("Cancel", true);
     GTEST_ASSERT_EQ(getTunnelFinishedFlag(mpp_tunnel_ptr), true);
     auto result_size = dynamic_cast<MockPacketWriter *>(writer_ptr.get())->write_packet_vec.size();
@@ -373,9 +379,7 @@ try
     std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockPacketWriter>();
     mpp_tunnel_ptr->connect(writer_ptr.get());
     GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->writeDone();
     GTEST_ASSERT_EQ(getTunnelFinishedFlag(mpp_tunnel_ptr), true);
     GTEST_ASSERT_EQ(dynamic_cast<MockPacketWriter *>(writer_ptr.get())->write_packet_vec.size(), 1);
@@ -390,9 +394,7 @@ try
     std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockPacketWriter>();
     mpp_tunnel_ptr->connect(writer_ptr.get());
     GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->getSyncTunnelSender()->consumerFinish("");
     waitSyncTunnelSenderThread(mpp_tunnel_ptr->getSyncTunnelSender());
 
@@ -410,9 +412,7 @@ TEST_F(TestMPPTunnel, WriteError)
         std::unique_ptr<PacketWriter> writer_ptr = std::make_unique<MockFailedWriter>();
         mpp_tunnel_ptr->connect(writer_ptr.get());
         GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
-        auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-        data_packet_ptr->getPacket().set_data("First");
-        mpp_tunnel_ptr->write(data_packet_ptr);
+        mpp_tunnel_ptr->write(newDataPacket("First"));
         mpp_tunnel_ptr->waitForFinish();
         GTEST_FAIL();
     }
@@ -433,9 +433,7 @@ TEST_F(TestMPPTunnel, WriteAfterFinished)
         mpp_tunnel_ptr->connect(writer_ptr.get());
         GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
         mpp_tunnel_ptr->close("Canceled", false);
-        auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-        data_packet_ptr->getPacket().set_data("First");
-        mpp_tunnel_ptr->write(data_packet_ptr);
+        mpp_tunnel_ptr->write(newDataPacket("First"));
         GTEST_FAIL();
     }
     catch (Exception & e)
@@ -504,9 +502,7 @@ try
     auto local_reader_ptr = connectLocalSyncTunnel(mpp_tunnel_ptr);
     GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
 
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->close("Cancel", false);
     local_reader_ptr->thread_manager->wait(); // Join local read thread
     GTEST_ASSERT_EQ(getTunnelSenderConsumerFinishedFlag(mpp_tunnel_ptr->getTunnelSender()), true);
@@ -523,9 +519,7 @@ try
     auto local_reader_ptr = connectLocalSyncTunnel(mpp_tunnel_ptr);
     GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
 
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->writeDone();
     local_reader_ptr->thread_manager->wait(); // Join local read thread
     GTEST_ASSERT_EQ(getTunnelSenderConsumerFinishedFlag(mpp_tunnel_ptr->getTunnelSender()), true);
@@ -542,9 +536,7 @@ try
     auto local_reader_ptr = connectLocalSyncTunnel(mpp_tunnel_ptr);
     GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
 
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->getTunnelSender()->consumerFinish("");
     local_reader_ptr->thread_manager->wait(); // Join local read thread
     GTEST_ASSERT_EQ(getTunnelSenderConsumerFinishedFlag(mpp_tunnel_ptr->getTunnelSender()), true);
@@ -561,9 +553,7 @@ TEST_F(TestMPPTunnel, LocalReadTerminate)
         mpp_tunnel_ptr->connect(nullptr);
         MockTerminateLocalReaderPtr local_reader_ptr = std::make_shared<MockTerminateLocalReader>(mpp_tunnel_ptr->getLocalTunnelSender());
         GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
-        auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-        data_packet_ptr->getPacket().set_data("First");
-        mpp_tunnel_ptr->write(data_packet_ptr);
+        mpp_tunnel_ptr->write(newDataPacket("First"));
         mpp_tunnel_ptr->waitForFinish();
         GTEST_FAIL();
     }
@@ -581,9 +571,7 @@ TEST_F(TestMPPTunnel, LocalWriteAfterFinished)
         auto local_reader_ptr = connectLocalSyncTunnel(mpp_tunnel_ptr);
         GTEST_ASSERT_EQ(getTunnelConnectedFlag(mpp_tunnel_ptr), true);
         mpp_tunnel_ptr->close("", false);
-        auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-        data_packet_ptr->getPacket().set_data("First");
-        mpp_tunnel_ptr->write(data_packet_ptr);
+        mpp_tunnel_ptr->write(newDataPacket("First"));
         mpp_tunnel_ptr->waitForFinish();
         GTEST_FAIL();
     }
@@ -605,11 +593,8 @@ try
 
     std::thread t(&MockAsyncCallData::run, call_data.get());
 
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
-    data_packet_ptr->getPacket().set_data("Second");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
+    mpp_tunnel_ptr->write(newDataPacket("Second"));
     mpp_tunnel_ptr->close("Cancel", true);
     GTEST_ASSERT_EQ(getTunnelFinishedFlag(mpp_tunnel_ptr), true);
 
@@ -631,9 +616,7 @@ try
 
     std::thread t(&MockAsyncCallData::run, call_data.get());
 
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->writeDone();
 
     GTEST_ASSERT_EQ(getTunnelFinishedFlag(mpp_tunnel_ptr), true);
@@ -654,9 +637,7 @@ try
 
     std::thread t(&MockAsyncCallData::run, call_data.get());
 
-    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-    data_packet_ptr->getPacket().set_data("First");
-    mpp_tunnel_ptr->write(data_packet_ptr);
+    mpp_tunnel_ptr->write(newDataPacket("First"));
     mpp_tunnel_ptr->getTunnelSender()->consumerFinish("");
     GTEST_ASSERT_EQ(getTunnelSenderConsumerFinishedFlag(mpp_tunnel_ptr->getTunnelSender()), true);
 
@@ -679,9 +660,7 @@ TEST_F(TestMPPTunnel, AsyncWriteError)
 
         std::thread t(&MockAsyncCallData::run, call_data.get());
 
-        auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
-        data_packet_ptr->getPacket().set_data("First");
-        mpp_tunnel_ptr->write(data_packet_ptr);
+        mpp_tunnel_ptr->write(newDataPacket("First"));
         t.join();
         mpp_tunnel_ptr->waitForFinish();
         GTEST_FAIL();

